### ISSUE TITLE
使用mpirun在多GPU环境下训练自定义数据集出现loss为nan得问题

### ISSUE THREAD

**Issue by Daniel-deng-yi** [mindspore-lab/mindyolo#250]
单卡训练自定义数据没问题，使用多GPU训练出现loss为nan问题
环境：Python 3.7.5
mindspore  2.0.0

启动命令：mpirun --allow-run-as-root -n 4 python train.py --config ./configs/yolov8/yolov8_m_test_djy.yaml  --is_parallel True

2023-11-28 08:21:16,093 [WARNING] overflow, still update, loss scale adjust to 1024.0
2023-11-28 08:21:16,219 [INFO] Epoch 1/500, Step 100/1048, imgsize (640, 640), loss: nan, lbox: nan, lcls: nan, dfl: nan, cur_lr: 0.09713740646839142
2023-11-28 08:21:16,223 [INFO] Epoch 1/500, Step 100/1048, step time: 4678.82 ms
2023-11-28 08:23:26,332 [WARNING] overflow, still update, loss scale adjust to 1024.0
2023-11-28 08:23:26,335 [INFO] Epoch 1/500, Step 200/1048, imgsize (640, 640), loss: nan, lbox: nan, lcls: nan, dfl: nan, cur_lr: 0.09427481144666672
2023-11-28 08:23:26,337 [INFO] Epoch 1/500, Step 200/1048, step time: 1301.13 ms
2023-11-28 08:25:36,584 [WARNING] overflow, still update, loss scale adjust to 1024.0
2023-11-28 08:25:36,587 [INFO] Epoch 1/500, Step 300/1048, imgsize (640, 640), loss: nan, lbox: nan, lcls: nan, dfl: nan, cur_lr: 0.09141221642494202
2023-11-28 08:25:36,589 [INFO] Epoch 1/500, Step 300/1048, step time: 1302.52 ms
2023-11-28 08:27:47,459 [WARNING] overflow, still update, loss scale adjust to 1024.0
2023-11-28 08:27:47,463 [INFO] Epoch 1/500, Step 400/1048, imgsize (640, 640), loss: nan, lbox: nan, lcls: nan, dfl: nan, cur_lr: 0.08854962140321732
2023-11-28 08:27:47,465 [INFO] Epoch 1/500, Step 400/1048, step time: 1308.76 ms
2023-11-28 08:29:58,440 [WARNING] overflow, still update, loss scale adjust to 1024.0
2023-11-28 08:29:58,444 [INFO] Epoch 1/500, Step 500/1048, imgsize (640, 640), loss: nan, lbox: nan, lcls: nan, dfl: nan, cur_lr: 0.08568702638149261
2023-11-28 08:29:58,446 [INFO] Epoch 1/500, Step 500/1048, step time: 1309.80 ms
2023-11-28 08:32:09,057 [WARNING] overflow, still update, loss scale adjust to 1024.0
2023-11-28 08:32:09,062 [INFO] Epoch 1/500, Step 600/1048, imgsize (640, 640), loss: nan, lbox: nan, lcls: nan, dfl: nan, cur_lr: 0.08282442390918732
2023-11-28 08:32:09,066 [INFO] Epoch 1/500, Step 600/1048, step time: 1306.20 ms
2023-11-28 08:34:19,511 [WARNING] overflow, still update, loss scale adjust to 1024.0
2023-11-28 08:34:19,516 [INFO] Epoch 1/500, Step 700/1048, imgsize (640, 640), loss: nan, lbox: nan, lcls: nan, dfl: nan, cur_lr: 0.07996182888746262
2023-11-28 08:34:19,519 [INFO] Epoch 1/500, Step 700/1048, step time: 1304.52 ms
2023-11-28 08:36:30,990 [WARNING] overflow, still update, loss scale adjust to 1024.0
2023-11-28 08:36:30,994 [INFO] Epoch 1/500, Step 800/1048, imgsize (640, 640), loss: nan, lbox: nan, lcls: nan, dfl: nan, cur_lr: 0.07709923386573792
2023-11-28 08:36:30,997 [INFO] Epoch 1/500, Step 800/1048, step time: 1314.76 ms
2023-11-28 08:38:41,955 [WARNING] overflow, still update, loss scale adjust to 1024.0
2023-11-28 08:38:41,958 [INFO] Epoch 1/500, Step 900/1048, imgsize (640, 640), loss: nan, lbox: nan, lcls: nan, dfl: nan, cur_lr: 0.07423663884401321
2023-11-28 08:38:41,960 [INFO] Epoch 1/500, Step 900/1048, step time: 1309.63 ms
2023-11-28 08:40:52,490 [WARNING] overflow, still update, loss scale adjust to 1024.0
2023-11-28 08:40:52,494 [INFO] Epoch 1/500, Step 1000/1048, imgsize (640, 640), loss: nan, lbox: nan, lcls: nan, dfl: nan, cur_lr: 0.07137404382228851
2023-11-28 08:40:52,496 [INFO] Epoch 1/500, Step 1000/1048, step time: 1305.35 ms

config文件配置如下：
optimizer:
  optimizer: momentum
  lr_init: 0.01  # initial learning rate (SGD=1E-2, Adam=1E-3)
  momentum: 0.937  # SGD momentum/Adam beta1
  nesterov: True  # update gradients with NAG(Nesterov Accelerated Gradient) algorithm
  loss_scale: 1.0  # loss scale for optimizer
  warmup_epochs: 3  # warmup epochs (fractions ok)
  warmup_momentum: 0.8  # warmup initial momentum
  warmup_bias_lr: 0.1  # warmup initial bias lr
  min_warmup_step: 1000  # minimum warmup step
  group_param: yolov8  # group param strategy
  gp_weight_decay: 0.0005  # group param weight decay 5e-4
  start_factor: 1.0
  end_factor: 0.01

loss:
  name: YOLOv8Loss
  box: 7.5  # box loss gain
  cls: 0.5  # cls loss gain
  dfl: 1.5  # dfl loss gain
  reg_max: 16

data:
  num_parallel_workers: 4
  # dataset_name: coco
  # train_set: /cache/coco/val2017.txt  # 118287 images
  # val_set: /cache/coco/val2017.txt  # 5000 images
  # test_set: /cache/coco/val2017.txt  # 20288 of 40670 images, submit to https://competitions.codalab.org/competitions/20794

  dataset_name: coco
  train_set: ./00.Datasets/train.json  # 118287 images
  val_set: ./00.Datasets/test.json # 5000 images
  test_set: ./00.Datasets/test.json

  nc: 2
  # class names
  names: ['call','call_doubt']
  # names: [ 'person', 'bicycle', 'car', 'motorcycle', 'airplane', 'bus', 'train', 'truck', 'boat', 'traffic light',
  #          'fire hydrant', 'stop sign', 'parking meter', 'bench', 'bird', 'cat', 'dog', 'horse', 'sheep', 'cow',
  #          'elephant', 'bear', 'zebra', 'giraffe', 'backpack', 'umbrella', 'handbag', 'tie', 'suitcase', 'frisbee',
  #          'skis', 'snowboard', 'sports ball', 'kite', 'baseball bat', 'baseball glove', 'skateboard', 'surfboard',
  #          'tennis racket', 'bottle', 'wine glass', 'cup', 'fork', 'knife', 'spoon', 'bowl', 'banana', 'apple',
  #          'sandwich', 'orange', 'broccoli', 'carrot', 'hot dog', 'pizza', 'donut', 'cake', 'chair', 'couch',
  #          'potted plant', 'bed', 'dining table', 'toilet', 'tv', 'laptop', 'mouse', 'remote', 'keyboard', 'cell phone',
  #          'microwave', 'oven', 'toaster', 'sink', 'refrigerator', 'book', 'clock', 'vase', 'scissors', 'teddy bear',
  #          'hair drier', 'toothbrush' ]

  # multi-stage data augment
  train_transforms: {
    stage_epochs: [ 490, 10 ],
    trans_list: [
      [
        { func_name: mosaic, prob: 1.0 },
        { func_name: copy_paste, prob: 0.1 },
        {func_name: resample_segments},
        { func_name: random_perspective, prob: 1.0, degrees: 0.0, translate: 0.1, scale: 0.9, shear: 0.0 },
        { func_name: mixup, alpha: 32.0, beta: 32.0, prob: 0.1, pre_transform: [
          { func_name: mosaic, prob: 1.0 },
          { func_name: copy_paste, prob: 0.1 },
          { func_name: resample_segments },
          { func_name: random_perspective, prob: 1.0, degrees: 0.0, translate: 0.1, scale: 0.9, shear: 0.0 }, ]
        },
        {func_name: albumentations},
        {func_name: hsv_augment, prob: 1.0, hgain: 0.015, sgain: 0.7, vgain: 0.4},
        {func_name: fliplr, prob: 0.5},
        {func_name: label_norm, xyxy2xywh_: True},
        {func_name: label_pad, padding_size: 160, padding_value: -1},
        {func_name: image_norm, scale: 255.},
        {func_name: image_transpose, bgr2rgb: True, hwc2chw: True}
      ],
      [
        {func_name: letterbox, scaleup: True},
        {func_name: resample_segments},
        {func_name: random_perspective, prob: 1.0, degrees: 0.0, translate: 0.1, scale: 0.9, shear: 0.0},
        {func_name: albumentations},
        {func_name: hsv_augment, prob: 1.0, hgain: 0.015, sgain: 0.7, vgain: 0.4},
        {func_name: fliplr, prob: 0.5},
        {func_name: label_norm, xyxy2xywh_: True},
        {func_name: label_pad, padding_size: 160, padding_value: -1},
        {func_name: image_norm, scale: 255.},
        {func_name: image_transpose, bgr2rgb: True, hwc2chw: True}
      ]]
  }

  test_transforms: [
    {func_name: letterbox, scaleup: False, only_image: True},
    {func_name: image_norm, scale: 255.},
    {func_name: image_transpose, bgr2rgb: True, hwc2chw: True}
  ]

epochs: 500  # total train epochs
per_batch_size: 16  # 16 * 8 = 128
img_size: 640
iou_thres: 0.7
conf_free: True
sync_bn: True
opencv_threads_num: 0  # opencv: disable threading optimizations

network:
  model_name: yolov8
  nc: 2  # number of classes
  reg_max: 16

  stride: [8, 16, 32]
  depth_multiple: 0.67  # scales module repeats
  width_multiple: 0.75  # scales convolution channels
  max_channels: 768

  # YOLOv8.0n backbone
  backbone:
    # [from, repeats, module, args]
    - [-1, 1, ConvNormAct, [64, 3, 2]]  # 0-P1/2
    - [-1, 1, ConvNormAct, [128, 3, 2]]  # 1-P2/4
    - [-1, 3, C2f, [128, True]]
    - [-1, 1, ConvNormAct, [256, 3, 2]]  # 3-P3/8
    - [-1, 6, C2f, [256, True]]
    - [-1, 1, ConvNormAct, [512, 3, 2]]  # 5-P4/16
    - [-1, 6, C2f, [512, True]]
    - [-1, 1, ConvNormAct, [1024, 3, 2]]  # 7-P5/32
    - [-1, 3, C2f, [1024, True]]
    - [-1, 1, SPPF, [1024, 5]]  # 9

  # YOLOv8.0n head
  head:
    - [-1, 1, Upsample, [None, 2, 'nearest']]
    - [[-1, 6], 1, Concat, [1]]  # cat backbone P4
    - [-1, 3, C2f, [512]]  # 12

    - [-1, 1, Upsample, [None, 2, 'nearest']]
    - [[-1, 4], 1, Concat, [1] ]  # cat backbone P3
    - [-1, 3, C2f, [256]]  # 15 (P3/8-small)

    - [-1, 1, ConvNormAct, [256, 3, 2]]
    - [[ -1, 12], 1, Concat, [1]]  # cat head P4
    - [-1, 3, C2f, [512]]  # 18 (P4/16-medium)

    - [-1, 1, ConvNormAct, [512, 3, 2]]
    - [[-1, 9], 1, Concat, [1]]  # cat head P5
    - [-1, 3, C2f, [1024]]  # 21 (P5/32-large)

    - [[15, 18, 21], 1, YOLOv8Head, [nc, reg_max, stride]]  # Detect(P3, P4, P5)

**Comment by zhanghuiyao** [comment]
当前mindyolo还不支持在gpu上训练，收敛情况不稳定

**Comment by zhanghuiyao** [comment]
这个问题可以尝试配置下每个step打印（--log_interval=1），看看第一个step是否出现nan，如果出现了可以排查下前向网络，看看哪个模块最开始出现的
